lib/Drush/Boot/DrupalBoot.php

```diff
@@ -2,115 +2,471 @@
 
 namespace Drush\Boot;
 
-class DrupalBoot implements Boot {
+abstract class DrupalBoot extends BaseBoot {
 
   function __construct() {
   }
 
-  // @todo We could split this into separate classes for Drupal 8, and Drupal 7 + 6,
-  // and give each one its own 'valid_root' implementation.  To do this, just add
-  // another candidate in drush_preflight_get_bootstrap_candidates().
   function valid_root($path) {
-    if (!empty($path) && is_dir($path) && file_exists($path . '/index.php')) {
-      // Drupal 8 root. Additional check for the presence of core/composer.json to
-      // grant it is not a Drupal 7 site with a base folder named "core".
-      $candidate = 'core/includes/common.inc';
-      if (file_exists($path . '/' . $candidate) && file_exists($path . '/core/core.services.yml')) {
-        if (file_exists($path . '/core/misc/drupal.js') || file_exists($path . '/core/assets/js/drupal.js')) {
-          return $candidate;
-        }
-      }
-      // Drupal 7 root.
-      $candidate = 'includes/common.inc';
-      if (file_exists($path . '/' . $candidate) && file_exists($path . '/misc/drupal.js')) {
-        return $candidate;
-      }
-    }
-    return FALSE;
   }
 
-  function preflight() {
-    // We need our constants before commandfile searching like DRUSH_BOOTSTRAP_DRUPAL_LOGIN.
-    require_once __DIR__ . '/bootstrap.inc';
-    require_once __DIR__ . '/command.inc';
+  function get_profile() {
+  }
+
+  function bootstrap_phases() {
+    return array(
+      DRUSH_BOOTSTRAP_DRUSH                  => 'bootstrap_drush',
+      DRUSH_BOOTSTRAP_DRUPAL_ROOT            => 'bootstrap_drupal_root',
+      DRUSH_BOOTSTRAP_DRUPAL_SITE            => 'bootstrap_drupal_site',
+      DRUSH_BOOTSTRAP_DRUPAL_CONFIGURATION   => 'bootstrap_drupal_configuration',
+      DRUSH_BOOTSTRAP_DRUPAL_DATABASE        => 'bootstrap_drupal_database',
+      DRUSH_BOOTSTRAP_DRUPAL_FULL            => 'bootstrap_drupal_full',
+      DRUSH_BOOTSTRAP_DRUPAL_LOGIN           => 'bootstrap_drupal_login');
+  }
+
+  function bootstrap_init_phases() {
+    return array(DRUSH_BOOTSTRAP_DRUSH, DRUSH_BOOTSTRAP_DRUPAL_FULL);
+  }
+
+  function enforce_requirement(&$command) {
+    parent::enforce_requirement($command);
+    $this->drush_enforce_requirement_drupal_dependencies($command);
+  }
+
+  function report_command_error($command) {
+    // If we reach this point, command doesn't fit requirements or we have not
+    // found either a valid or matching command.
 
-    drush_set_context('DRUSH_BOOTSTRAP_PHASE', DRUSH_BOOTSTRAP_NONE);
+    // If no command was found check if it belongs to a disabled module.
+    if (!$command) {
+      $command = $this->drush_command_belongs_to_disabled_module();
+    }
+    parent::report_command_error($command);
   }
 
   function command_defaults() {
     return array(
+      'drupal dependencies' => array(),
       'bootstrap' => DRUSH_BOOTSTRAP_DRUPAL_LOGIN,
```

Comment:
It occurs to me that the fact that we do LOGIN by default is a holdover from ancient spaghetti Drupal. We might be able to change that for D7/D8. Something to consider for the future.